packages/sdk/src/fee/FeeEstimator.ts

```diff
@@ -0,0 +1,144 @@
+import { Psbt } from "bitcoinjs-lib"
+
+import { AddressFormats, getNetwork, getScriptType } from ".."
+import { Network } from "../config/types"
+import { MAXIMUM_FEE } from "../constants"
+import { FeeEstimatorOptions } from "./types"
+
+export default class FeeEstimator {
+  feeRate: number
+  network: Network
+  psbt!: Psbt
+  witness?: Buffer[] = []
+  fee = 0
+
+  private virtualSize = 0
+  private weight = 0
+
+  constructor({ feeRate, network, psbt, witness }: FeeEstimatorOptions) {
+    if (feeRate < 0 || !Number.isSafeInteger(feeRate)) {
+      throw new Error("Invalid feeRate")
+    }
+
+    this.feeRate = +feeRate // convert decimal to whole number that might have passed Number.isSafeInteger check due to precision loss
+    this.network = network
+    this.witness = witness || []
+    this.psbt = psbt || new Psbt({ network: getNetwork(this.network) })
+  }
+
+  private sanityCheckFee() {
+    if (this.fee > MAXIMUM_FEE) {
+      throw new Error("Error while calculating fees")
+    }
+  }
+
+  calculateNetworkFee(): number {
+    this.fee = this.calculateVirtualSize() * this.feeRate
+    this.sanityCheckFee()
+
+    return this.fee
+  }
+
+  private analyzePSBTComponents() {
+    const inputs = this.psbt.data.inputs
+    const outputs = this.psbt.txOutputs
+    const inputTypes: AddressFormats[] = []
+    const outputTypes: AddressFormats[] = []
+
+    if (inputs.length === 0) {
+      throw new Error("PSBT must have at least one input")
+    }
+
+    if (outputs.length === 0) {
+      throw new Error("PSBT must have at least one output")
+    }
+
+    inputs.forEach((input) => {
+      const script = input.witnessUtxo && input.witnessUtxo.script ? input.witnessUtxo.script : null
+
+      if (!script) {
+        throw new Error("Invalid script")
+      }
+
+      inputTypes.push(getScriptType(script, this.network))
+    })
+
+    outputs.forEach((output) => {
+      outputTypes.push(getScriptType(output.script, this.network))
+    })
+
+    return {
+      inputTypes,
+      outputTypes
+    }
+  }
+
+  private calculateScriptWitnessSize() {
+    return this.analyzePSBTComponents().inputTypes.includes("taproot") && this.witness?.length
+      ? this.witness.reduce((acc, witness) => (acc += witness.byteLength), 0) || 0
+      : 0
+  }
+
+  private getBaseSize() {
+    const { inputTypes, outputTypes } = this.analyzePSBTComponents()
+    const witnessHeaderSize = 2
+    const inputVBytes = inputTypes.reduce(
+      (acc, inputType) => {
+        const { input, txHeader, witness } = this.getBaseSizeByType(inputType)
+        acc.txHeader = txHeader
```

Comment:
is this supposed to be `+=` as well?

Reply:
Nope. There always 1 header in a tx.

Reply:
but this means that the header will be based on the last input type?
so if there's 3 inputs of legacy,legacy and taproot, the header will be of taproot?